File: 121. Best Time to Buy and Sell Stock/maxProfit.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        int maxProfit(vector<int>& prices) {
            int len = prices.size();
            int min_price = prices[0];
            vector<int> dp(len , 0);
    
            for(int i=1 ; i<len; i++){
                min_price = min(min_price, prices[i]);
                dp[i] = max(dp[i-1] , prices[i] - min_price);
            }
    
            return dp[len-1];
        }
    };

File: 121. Best Time to Buy and Sell Stock/maxProfit.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
    let len = prices.length
    let min = prices[0]
    let dp = new Array(len).fill(0)

    for (let i = 1; i < len; i++) {
        min = Math.min(min, prices[i])
        //which is GREATER
        // the previous PROFIT
        //or if i SELL today will get the MAX PROFIT ???
        dp[i] = Math.max(dp[i - 1], prices[i] - min)
    }

    return dp[len - 1]
};

///SOLUTION 2


/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
    let len = prices.length
    let min = prices[0]
    let maxProfit = 0

    for (let i = 1; i < len; i++) {
        min = Math.min(prices[i], min) //buy
        let curr = prices[i] - min  //sell
        maxProfit = Math.max(curr, maxProfit)
    }

    return maxProfit
};